### DISCUSSION TITLE
FastAPI in Docker: Httpx request to external API 500 + timeout

### DISCUSSION THREAD

**Discussion by fbpyr** [fastapi/fastapi#6655] [Questions] [answered]
### Example

```Python
from fastapi import FastAPI, APIRouter
import httpx


app = FastAPI(root_path="/projects")


@app.get("/get_req_httpx/", response_description="Async request with httpx")
async def get_projects():
    projects = []
    url = "http://httpbin.org/get"
    async with httpx.AsyncClient() as client:
        resp = await client.get(url)
        print(resp)
        projects = resp.json()
    return {"projects": projects}
```

### Description

First of all thank you so much for awesome FastAPI!!

I wrote a few smaller projects with it and enjoeyd it a a lot, but this is the first 
time I need to contact an external API, which I tried with Httpx . 
Unfortunately it is not clear to me, if this is on Httpx or FastAPI or Docker - 
feel free to redirect me if it is/gets clear.

When I run the above code on my local machine without Docker, it works fine. 
If I curl or use synchronous Httpx python to request the above content inside 
the docker container it also works.
But as soon as the above minimal part of FastAPI app runs in Docker I always 
get a 500 server error with the traceback below.
Same with aiohttp (async) or requests (sync)

Thank you in advance. 🙂 

* Open the browser and call the endpoint `/projects/get_req_httpx/`.
* It returns a JSON with gets a internal 500 server error with httpx.ConnectTimeout.
* But I expected it to return `{"projects": "all the info from httpbin.org/get args, headers, etc.."}`..

### Environment

* OS: Linux Docker Container on Debian 10 buster host
* Python version: Python 3.8.2
* FastAPI Version 0.62.0 (but same error with 0.65.1)
* Httpx Version 0.18.1

### Additional context

From Dockerfile:
```
FROM tiangolo/uvicorn-gunicorn-fastapi:python3.8
...
RUN python3 -m pip install httpx
```

The traceback:
```bash
172.27.0.2:42744 - "GET /projects/docs HTTP/1.1" 200
172.27.0.2:42744 - "GET /projects/openapi.json HTTP/1.1" 200
172.27.0.2:42744 - "GET /projects/api_v1/projects/get_req_httpx/ HTTP/1.1" 500
[2021-05-21 xx:xx:xx +0000] [16] [ERROR] Exception in ASGI application
 Traceback (most recent call last):
   File "/usr/local/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 385, in run_asgi
     result = await app(self.scope, self.receive, self.send)
   File "/usr/local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
     return await self.app(scope, receive, send)
   File "/usr/local/lib/python3.8/site-packages/fastapi/applications.py", line 190, in __call__
   await super().__call__(scope, receive, send)
   File "/usr/local/lib/python3.8/site-packages/starlette/applications.py", line 111, in __call__
     await self.middleware_stack(scope, receive, send)
   File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
     raise exc from None
   File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
     await self.app(scope, receive, _send)
   File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
     raise exc from None
   File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
     await self.app(scope, receive, sender)
   File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 566, in __call__
     await route.handle(scope, receive, send)
   File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 227, in handle
     await self.app(scope, receive, send)
   File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 41, in app
     response = await func(request)
   File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 188, in app
     raw_response = await run_endpoint_function(
   File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 135, in run_endpoint_function
     return await dependant.call(**values)
   File "/app/app/server/app.py", line 14, in get_projects
     resp = await client.get(url)
   File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1661, in get
     return await self.request(
   File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1425, in request
     response = await self.send(
   File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1510, in send
     response = await self._send_handling_auth(
   File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1545, in _send_handling_auth
     response = await self._send_handling_redirects(
   File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1579, in _send_handling_redirects
     response = await self._send_single_request(request, timeout)
   File "/usr/local/lib/python3.8/site-packages/httpx/_client.py", line 1620, in _send_single_request
     ) = await transport.handle_async_request(
   File "/usr/local/lib/python3.8/site-packages/httpx/_transports/default.py", line 274, in handle_async_request
     (
   File "/usr/local/lib/python3.8/contextlib.py", line 131, in __exit__
     self.gen.throw(type, value, traceback)
   File "/usr/local/lib/python3.8/site-packages/httpx/_transports/default.py", line 78, in map_httpcore_exceptions
     raise mapped_exc(message) from exc
httpx.ConnectTimeout
Future exception was never retrieved
future: <Future finished exception=gaierror(-3, 'Temporary failure in name resolution')>
socket.gaierror: [Errno -3] Temporary failure in name resolution
```

**Comment by ArcLightSlavik** [comment]
How are you starting the image?

Both of examples below worked for me
```
docker run -d --name fastapi_test_1 -p 80:80 fastapi_test
```

```
version: '3'

services:

  testing:
    image: fastapi_testing:0.0.1
    build:
      context: .
      dockerfile: Dockerfile
    ports:
      - "1234:1234"
    environment:
      PORT: 1234
    volumes:
      - ./app:/home/app
```

**Comment by fbpyr** [comment]
@ArcLightSlavik thank you so much for trying it out. 
I usually start the it with docker compose: 
`docker-compose -f docker-compose.yml up --build` 
as it sits behind traefik, so I pass the traefik labels 
to the fastapi container in the docker-compose.

Weirdly enough the above runs without problems on my raspi at home. 
So I thought it might be because of how I build it. 
But even with a fresh Dockerfile:
```Dockerfile
FROM python:3.8

RUN python3 -m pip install fastapi uvicorn httpx

COPY ./app /app

CMD ["uvicorn", "app.main:app", "--host", "0.0.0.0", "--port", "80"]
```
the error/timeout persists.

I wonder if it is related to: 
https://github.com/tiangolo/fastapi/issues/3137#issuecomment-833886718 
as I get the same strange behaviour: 
I do not get the 500 right away but only after a couple of seconds. 🤔 

When using httpx I get the aforementioned timeout, 
if I use requests to send the request I get: 
`Failed to establish a new connection: [Errno -3] Temporary failure in name resolution`

From within that very container, if I request it from a separate python3 interpreter:
```python
import httpx

with httpx.Client() as cl:
    resp = cl.get("http://httpbin.org/get")

print(resp)  # gives me <Response [200 OK]>
```
the httpx request works fine and as expected, which makes me doubt it 
that the name resolution would be the problem.

**Comment by ArcLightSlavik** [comment]
My guess would be is that it's something related to traefik, this perhaps https://github.com/nextcloud/docker/issues/1320.

**Comment by fbpyr** [comment]
@ArcLightSlavik thank you for the link. 
If traefik was blocking the traffic, should not curl, ping 
and python3 request from within the container also 
be unable to connect? (but they are able)

Maybe it has to do with uvicorn.(?)

I added a simple ping function:
```python
from fastapi import FastAPI, APIRouter
import httpx
import asyncio


app = FastAPI(root_path="/projects")


@app.get("/get_req_httpx/", response_description="Async request with httpx")
async def get_projects():
    projects = []
    url = "http://httpbin.org/get"
    async with httpx.AsyncClient() as client:
        resp = await client.get(url)
        print(resp)
        projects = resp.json()
    return {"projects": projects}


@app.get("/get_ping/", response_description="Async subproc ping")
async def get_ping():
    ping_proc = await asyncio.create_subprocess_shell(
        "ping -c2 localhost", # localhost works, other IPs do not.
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, stderr = await ping_proc.communicate()
    print(f"INFO: ping: {stdout =} {stderr =}")
    return {"ping": f"{stdout =} {stderr =}"}

```
which interestingly can ping localhost, but no external IPs. 
( Which explains why it cannot resolve names. )

Why does the main entrypoint ( uvicorn+fastapi+httpx/aiohttp/requests ) 
seem like it is shielded from net access, when other applications in 
the same container (curl, ping, python3 with requests, etc..) are not? 😕

**Comment by ricky-sb** [comment]
@fbpyr, maybe try setting your DNS explicitly in the `docker-compose.yml` file?  

```yml
services:
  web:
    dns:
    - 8.8.8.8
    - 8.8.4.4
 ```

**Comment by fbpyr** [comment, marked as answer]
I finally figured it out: 🎉 

I got it to work with a docker image, based on newer python:3.9.5-slim , 
where everything works now as expected. 🙂 

But the main problem was actually, that in the docker-compose, 
I had only included the container to be part of a network, 
that had no net access, but kept jumping between probing the 
docker container with either directly docker: 
`docker run -it {name} /bin/bash` 
where net access worked fine, 
or connecting to the container already started with docker-compose, 
where no net access worked:
`docker exec -it {container_id} /bin/bash` 

@ArcLightSlavik @ricky-sb thank you so much the support! 👍

**Comment by jomadhuri** [comment]
Hi @fbpyr can you please direct me to the changes you made in the docker-compose file to make it work?
